casper/src/main/scala/coop/rchain/casper/util/rholang/InterpreterUtil.scala

```diff
@@ -86,67 +61,91 @@ object InterpreterUtil {
-                none[StateHash].asRight[BlockError].pure
-            case ReplayStatusMismatch(replay: DeployStatus, orig: DeployStatus) =>
-              Log[F].warn(
-                s"Found replay status mismatch; replay failure is ${replay.isFailed} and orig failure is ${orig.isFailed}"
-              ) >>
-                none[StateHash].asRight[BlockError].pure
-            case UnknownFailure =>
-              Log[F].warn(s"Found unknown failure") >>
-                none[StateHash].asRight[BlockError].pure
-            case UnusedCommEvent(_) =>
-              Sync[F].raiseError(new RuntimeException("found UnusedCommEvent"))
-          }
-        case Left((None, status)) =>
-          status match {
-            case UnusedCommEvent(_: ReplayException) =>
+    result.pure.flatMap {
+      case Left((Some(deploy), status)) =>
+        status match {
+          case InternalErrors(exs) =>
+            BlockStatus
+              .exception(
+                new Exception(s"Internal errors encountered while processing ${PrettyPrinter
+                  .buildString(deploy)}: ${exs.mkString("\n")}")
+              )
+              .asLeft[Option[StateHash]]
+              .pure
+          case UserErrors(errors: Seq[Throwable]) =>
+            Log[F].warn(s"Found user error(s) ${errors.map(_.getMessage).mkString("\n")}") >>
               none[StateHash].asRight[BlockError].pure
-            case InternalErrors(_) => new RuntimeException("found InternalErrors").raiseError
-            case ReplayStatusMismatch(_, _) =>
-              new RuntimeException("found ReplayStatusMismatch").raiseError
-            case UnknownFailure => new RuntimeException("found UnknownFailure").raiseError
-            case UserErrors(_)  => new RuntimeException("found UserErrors").raiseError
-          }
-        case Right(computedStateHash) =>
-          if (tsHash == computedStateHash) {
-            // state hash in block matches computed hash!
-            computedStateHash.some.asRight[BlockError].pure
-          } else {
-            // state hash in block does not match computed hash -- invalid!
-            // return no state hash, do not update the state hash set
+          case ReplayStatusMismatch(replay: DeployStatus, orig: DeployStatus) =>
             Log[F].warn(
-              s"Tuplespace hash ${PrettyPrinter.buildString(tsHash)} does not match computed hash ${PrettyPrinter
-                .buildString(computedStateHash)}."
+              s"Found replay status mismatch; replay failure is ${replay.isFailed} and orig failure is ${orig.isFailed}"
             ) >>
               none[StateHash].asRight[BlockError].pure
-
-          }
-      }
+          case UnknownFailure =>
+            Log[F].warn(s"Found unknown failure") >>
+              none[StateHash].asRight[BlockError].pure
+          case UnusedCommEvent(_) =>
+            Sync[F].raiseError(new RuntimeException("found UnusedCommEvent"))
```

Comment:
While you're in the area, it's a good idea to shorten this as well.

Reply:
It looks like receiving a `(Some, UnusedCommEvent)` pair is supposed to be indicative of an error a level above the validation protocol. It looks like it's supposed to represent an error. If so, `BugFoundError` should be used here.

Reply:
Done :)